injection/sharedmain/main.go

```diff
@@ -424,30 +469,24 @@ func WatchLoggingConfigOrDie(ctx context.Context, cmw *cminformer.InformedWatche
 // WatchObservabilityConfigOrDie establishes a watch of the observability config
 // or dies by calling log.Fatalw. Note, if the config does not exist, it will be
 // defaulted and this method will not die.
-func WatchObservabilityConfigOrDie(ctx context.Context, cmw *cminformer.InformedWatcher, profilingHandler *profiling.Handler, logger *zap.SugaredLogger, component string) {
-	if _, err := kubeclient.Get(ctx).CoreV1().ConfigMaps(system.Namespace()).Get(ctx, metrics.ConfigMapName(),
-		metav1.GetOptions{}); err == nil {
-		cmw.Watch(metrics.ConfigMapName(),
-			metrics.ConfigMapWatcher(ctx, component, SecretFetcher(ctx), logger),
-			profilingHandler.UpdateFromConfigMap)
-	} else if !apierrors.IsNotFound(err) {
-		logger.Fatalw("Error reading ConfigMap "+metrics.ConfigMapName(), zap.Error(err))
+func WatchObservabilityConfigOrDie(
+	ctx context.Context,
+	cmw *cminformer.InformedWatcher,
+	pprof *pprofServer,
+	logger *zap.SugaredLogger,
+	component string,
+) {
+	cmName := o11yconfigmap.Name()
+	client := kubeclient.Get(ctx).CoreV1().ConfigMaps(system.Namespace())
+
+	observers := []configmap.Observer{
+		pprof.UpdateFromConfigMap,
 	}
-}
 
-// SecretFetcher provides a helper function to fetch individual Kubernetes
-// Secrets (for example, a key for client-side TLS). Note that this is not
-// intended for high-volume usage; the current use is when establishing a
-// metrics client connection in WatchObservabilityConfigOrDie.
-func SecretFetcher(ctx context.Context) metrics.SecretFetcher {
-	// NOTE: Do not use secrets.Get(ctx) here to get a SecretLister, as it will register
-	// a *global* SecretInformer and require cluster-level `secrets.list` permission,
-	// even if you scope down the Lister to a given namespace after requesting it. Instead,
-	// we package up a function from kubeclient.
-	// TODO(evankanderson): If this direct request to the apiserver on each TLS connection
-	// to the opencensus agent is too much load, switch to a cached Secret.
-	return func(name string) (*corev1.Secret, error) {
-		return kubeclient.Get(ctx).CoreV1().Secrets(system.Namespace()).Get(ctx, name, metav1.GetOptions{})
+	if _, err := client.Get(ctx, cmName, metav1.GetOptions{}); err == nil {
+		cmw.Watch(cmName, observers...)
+	} else if !apierrors.IsNotFound(err) {
+		logger.Fatalw("Error reading ConfigMap "+cmName, zap.Error(err))
```

Comment:
I'm assuming we don't have a way to notice if the ConfigMap shows up later?

Reply:
This file could use a cleanup - I've just changed the 'Watch' here but the diff shows a larger change.

Generally, I think the controllers should _halt_ and wait if a config map is not present. I could imagine a split brain scenario occurring. 

I'd rather not make those edits in this PR.